### PR TITLE
Elastic v2 support and elastic expr refactor

### PR DESCRIPTION
**Pull request by kylebrandt**

### CODE REVIEW

**Comment by kylebrandt** [comment]
Still needs documentation, other than that I think it is ready to go. Will do that Monday.

**Comment by kylebrandt** [comment]
```2016/01/19 09:44:27 info: check.go:256: hardware{host=ny-sql04}: template: header:22:48: executing "header" at <.Last.IncidentId>: IncidentId is not a field of struct type models.Event
2016/01/19 09:44:27 info: check.go:269: hardware{host=ny-sql04}: template: header:22:48: executing "header" at <.Last.IncidentId>: IncidentId is not a field of struct type models.Event```

File: docs/expressions.md [hunk 1]
```diff
@@ -176,6 +178,68 @@ lstat returns various summary stats per bucket for the specified `field`. The fi
   * As of January 15, 2015 - logstash functionality is new so these functions may change a fair amount based on experience using them in alerts.
   * Alerts using this information likely want to set ignoreUnknown, since only "groups" that appear in the time frame are in the results.
 
+## Elastic Query Functions
+
+Elasitc replaces the deprecated logstash (ls) functions. It only works with Elastic v2+. It introduces two new types to allow for greater flexibility in querying. The ESIndexer type generates indexes to query (based on the date range). There are now different functions to generate indexers for people with different configurations. The ESQuery type is generates elastic queries so you can filter your results. By making these new types, new Indexers and Elastic queries can be added over time.
+
```
Comment (Dieterbe) [inline docs/expressions.md]: generate indexes? do you mean generate index names [for which indexes to be queried] ?

Comment (kylebrandt) [inline docs/expressions.md]: Yes, index names, will update. Thanks!

File: docs/expressions.md [hunk 2]
```diff
@@ -176,6 +178,68 @@ lstat returns various summary stats per bucket for the specified `field`. The fi
   * As of January 15, 2015 - logstash functionality is new so these functions may change a fair amount based on experience using them in alerts.
   * Alerts using this information likely want to set ignoreUnknown, since only "groups" that appear in the time frame are in the results.
 
+## Elastic Query Functions
+
+Elasitc replaces the deprecated logstash (ls) functions. It only works with Elastic v2+. It introduces two new types to allow for greater flexibility in querying. The ESIndexer type generates indexes to query (based on the date range). There are now different functions to generate indexers for people with different configurations. The ESQuery type is generates elastic queries so you can filter your results. By making these new types, new Indexers and Elastic queries can be added over time.
```
Comment (Dieterbe) [inline docs/expressions.md]: so now we can also use these ES functions for any ES querying, without being tied to particular logstash schema? can this be clarified? are there any known limitations that we may want to document?

Comment (kylebrandt) [inline docs/expressions.md]: @Dieterbe  Right, this meant to make it so we don't tie to a particular schema. There are elastic queries beyond what I created functions for, but if people request them they can just be new funcs that return ESQuery. Same goes for the index naming schema. 

Regexes might not all make it through the parser - there is an open issue for that. But I think that would be fixed independently of this.

One limitation I will add is that if your fields is analyzed, you get multiple histograms based on the terms. But turning an analyzed field into a tag key doesn't really make sense in any scenario I can think of.

**Comment by Dieterbe** [comment]
Nice work Kyle! Thanks.
Looking forward to play with this at some point.

File: docs/expressions.md (hunk 2 above)
Comment (kylebrandt) [inline docs/expressions.md]: @Dieterbe Ah you made me think of one:

`elastic.NewDateHistogramAggregation().Field("@timestamp").`

`@timestamp` is hardcoded, maybe I will make that an argument for the indexer...

**Comment by deanefrati** [comment]
this looks great @kylebrandt! Can't wait to try it (hopefully this wee). Quick question are you planning to handle the scenario mentioned in https://github.com/bosun-monitor/bosun/issues/1160 as part of this work?

**Comment by kylebrandt** [comment]
@deanefrati I looked at this and didn't want to do it in this PR, but will make sure it is on my radar for soon. I think that issue is impacting us at stack well to make elastic usable for alerts. I just had enough to wrap my head around with elastic's API changing, the elastic go library decided that was a good time to do some refactoring since elastic changed, and then I decided it was a good time to do it in bosun too. That is about as much as I could hold in my head :-P

**Comment by deanefrati** [comment]
makes total sense. thanks again

File: cmd/bosun/conf/conf.go [hunk 3]
```diff
@@ -67,6 +67,7 @@ type Conf struct {
 	GraphiteHost         string                    // Graphite query host: foo.bar.baz
 	GraphiteHeaders      []string                  // extra http headers when querying graphite.
 	LogstashElasticHosts expr.LogstashElasticHosts // CSV Elastic Hosts (All part of the same cluster) that stores logstash documents, i.e http://ny-elastic01:9200
+	ElasticHosts         expr.ElasticHosts         // CSV Elastic Hosts (All part of the same cluster), i.e http://ny-elastic01:9200
```
Comment (captncraig) [inline cmd/bosun/conf/conf.go]: I would like this to be more clear which is for v1 and which is for v2

Comment (kylebrandt) [inline cmd/bosun/conf/conf.go]: @captncraig updated comments and the documentation.

File: cmd/bosun/expr/elastic.go [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+	"esindices": {
+		Args:     []parse.FuncType{parse.TypeString, parse.TypeString},
+		VArgs:    true,
+		VArgsPos: 1,
+		Return:   parse.TypeESIndexer,
+		F:        ESIndicies,
+	},
+	"esdaily": {
+		Args:     []parse.FuncType{parse.TypeString, parse.TypeString, parse.TypeString},
+		VArgsPos: 1,
```
Comment (captncraig) [inline cmd/bosun/expr/elastic.go]: VArgs: false and VargsPos: 1? I'm not sure what that combination means.

Comment (kylebrandt) [inline cmd/bosun/expr/elastic.go]: Is a mistake

**Pull request merged**